// @alfresco/adf-process-services/esm2015/people/components/people-list/people-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { DataTableComponent } from '@alfresco/adf-core';
import { DataColumnListComponent } from '@alfresco/adf-core';
import { Component, ContentChild, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { UserEventModel } from '../../../task-list/models/user-event.model';
export class PeopleListComponent {
    constructor() {
        /**
         * Toggles whether or not actions should be visible, i.e. the 'Three-Dots' menu.
         */
        this.actions = false;
        /**
         * Emitted when the user clicks a row in the people list.
         */
        this.clickRow = new EventEmitter();
        /**
         * Emitted when the user clicks in the 'Three Dots' drop down menu for a row.
         */
        this.clickAction = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.peopleDataTable.columnList = this.columnList;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
    }
    /**
     * @param {?} event
     * @return {?}
     */
    selectUser(event) {
        this.user = event.value.obj;
        this.clickRow.emit(this.user);
    }
    /**
     * @return {?}
     */
    hasActions() {
        return this.actions;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onShowRowActionsMenu(event) {
        /** @type {?} */
        let removeAction = {
            title: 'Remove',
            name: 'remove'
        };
        event.value.actions = [
            removeAction
        ];
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onExecuteRowAction(event) {
        /** @type {?} */
        let args = event.value;
        /** @type {?} */
        let action = args.action;
        this.clickAction.emit(new UserEventModel({ type: action.name, value: args.row.obj }));
    }
}
PeopleListComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-people-list',
                template: "<adf-datatable\n    #dataTable\n    class=\"adf-task-people-list\"\n    [rows]=\"users\"\n    [actions]=\"hasActions()\"\n    [showHeader]=\"false\"\n    (rowClick)=\"selectUser($event)\"\n    (rowDblClick)=\"selectUser($event)\"\n    (showRowActionsMenu)=\"onShowRowActionsMenu($event)\"\n    (executeRowAction)=\"onExecuteRowAction($event)\">\n</adf-datatable>\n",
                styles: [""]
            }] }
];
PeopleListComponent.propDecorators = {
    columnList: [{ type: ContentChild, args: [DataColumnListComponent,] }],
    peopleDataTable: [{ type: ViewChild, args: ['dataTable',] }],
    users: [{ type: Input }],
    actions: [{ type: Input }],
    clickRow: [{ type: Output }],
    clickAction: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PeopleListComponent.prototype.columnList;
    /** @type {?} */
    PeopleListComponent.prototype.peopleDataTable;
    /**
     * The array of user data used to populate the people list.
     * @type {?}
     */
    PeopleListComponent.prototype.users;
    /**
     * Toggles whether or not actions should be visible, i.e. the 'Three-Dots' menu.
     * @type {?}
     */
    PeopleListComponent.prototype.actions;
    /**
     * Emitted when the user clicks a row in the people list.
     * @type {?}
     */
    PeopleListComponent.prototype.clickRow;
    /**
     * Emitted when the user clicks in the 'Three Dots' drop down menu for a row.
     * @type {?}
     */
    PeopleListComponent.prototype.clickAction;
    /** @type {?} */
    PeopleListComponent.prototype.user;
}
