// @alfresco/adf-core/esm5/viewer/components/txtViewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { HttpClient } from '@angular/common/http';
import { Component, Input, ViewEncapsulation } from '@angular/core';
var TxtViewerComponent = /** @class */ (function () {
    function TxtViewerComponent(http) {
        this.http = http;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    TxtViewerComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var blobFile = changes['blobFile'];
        if (blobFile && blobFile.currentValue) {
            return this.readBlob(blobFile.currentValue);
        }
        /** @type {?} */
        var urlFile = changes['urlFile'];
        if (urlFile && urlFile.currentValue) {
            return this.getUrlContent(urlFile.currentValue);
        }
        if (!this.urlFile && !this.blobFile) {
            throw new Error('Attribute urlFile or blobFile is required');
        }
    };
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    TxtViewerComponent.prototype.getUrlContent = /**
     * @private
     * @param {?} url
     * @return {?}
     */
    function (url) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.http.get(url, { responseType: 'text' }).subscribe(function (res) {
                _this.content = res;
                resolve();
            }, function (event) {
                reject(event);
            });
        });
    };
    /**
     * @private
     * @param {?} blob
     * @return {?}
     */
    TxtViewerComponent.prototype.readBlob = /**
     * @private
     * @param {?} blob
     * @return {?}
     */
    function (blob) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            /** @type {?} */
            var reader = new FileReader();
            reader.onload = function () {
                _this.content = reader.result;
                resolve();
            };
            reader.onerror = function (error) {
                reject(error);
            };
            reader.readAsText(blob);
        });
    };
    TxtViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-txt-viewer',
                    template: "<pre class=\"adf-txt-viewer-content\">\n    {{content}}\n</pre>\n",
                    host: { 'class': 'adf-txt-viewer' },
                    encapsulation: ViewEncapsulation.None,
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TxtViewerComponent.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    TxtViewerComponent.propDecorators = {
        urlFile: [{ type: Input }],
        blobFile: [{ type: Input }]
    };
    return TxtViewerComponent;
}());
export { TxtViewerComponent };
if (false) {
    /** @type {?} */
    TxtViewerComponent.prototype.urlFile;
    /** @type {?} */
    TxtViewerComponent.prototype.blobFile;
    /** @type {?} */
    TxtViewerComponent.prototype.content;
    /**
     * @type {?}
     * @private
     */
    TxtViewerComponent.prototype.http;
}
