// @alfresco/adf-core/esm5/services/auth-guard-bpm.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AppConfigService, AppConfigValues } from '../app-config/app-config.service';
import { AuthenticationService } from './authentication.service';
import * as i0 from "@angular/core";
import * as i1 from "./authentication.service";
import * as i2 from "@angular/router";
import * as i3 from "../app-config/app-config.service";
var AuthGuardBpm = /** @class */ (function () {
    function AuthGuardBpm(authService, router, appConfigService) {
        this.authService = authService;
        this.router = router;
        this.appConfigService = appConfigService;
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardBpm.prototype.canActivate = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        return this.checkLogin(state.url);
    };
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardBpm.prototype.canActivateChild = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        return this.canActivate(route, state);
    };
    /**
     * @param {?} redirectUrl
     * @return {?}
     */
    AuthGuardBpm.prototype.checkLogin = /**
     * @param {?} redirectUrl
     * @return {?}
     */
    function (redirectUrl) {
        /** @type {?} */
        var withCredentialsMode = this.appConfigService.get('auth.withCredentials', false);
        if (this.authService.isBpmLoggedIn() || withCredentialsMode) {
            return true;
        }
        if (!this.authService.isOauth() || this.isOAuthWithoutSilentLogin()) {
            this.authService.setRedirect({ provider: 'BPM', url: redirectUrl });
            /** @type {?} */
            var pathToLogin = this.getRouteDestinationForLogin();
            this.router.navigate(['/' + pathToLogin]);
        }
        return false;
    };
    /**
     * @return {?}
     */
    AuthGuardBpm.prototype.isOAuthWithoutSilentLogin = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var oauth = this.appConfigService.get(AppConfigValues.OAUTHCONFIG, null);
        return this.authService.isOauth() && oauth.silentLogin === false;
    };
    /**
     * @private
     * @return {?}
     */
    AuthGuardBpm.prototype.getRouteDestinationForLogin = /**
     * @private
     * @return {?}
     */
    function () {
        return this.appConfigService && this.appConfigService.get(AppConfigValues.LOGIN_ROUTE) ? this.appConfigService.get(AppConfigValues.LOGIN_ROUTE) : 'login';
    };
    AuthGuardBpm.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AuthGuardBpm.ctorParameters = function () { return [
        { type: AuthenticationService },
        { type: Router },
        { type: AppConfigService }
    ]; };
    /** @nocollapse */ AuthGuardBpm.ngInjectableDef = i0.defineInjectable({ factory: function AuthGuardBpm_Factory() { return new AuthGuardBpm(i0.inject(i1.AuthenticationService), i0.inject(i2.Router), i0.inject(i3.AppConfigService)); }, token: AuthGuardBpm, providedIn: "root" });
    return AuthGuardBpm;
}());
export { AuthGuardBpm };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthGuardBpm.prototype.authService;
    /**
     * @type {?}
     * @private
     */
    AuthGuardBpm.prototype.router;
    /**
     * @type {?}
     * @private
     */
    AuthGuardBpm.prototype.appConfigService;
}
