// @alfresco/adf-core/esm5/services/external-alfresco-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { AlfrescoApiCompatibility } from '@alfresco/js-api';
import * as i0 from "@angular/core";
/* tslint:disable:adf-file-name */
var ExternalAlfrescoApiService = /** @class */ (function () {
    function ExternalAlfrescoApiService() {
    }
    /**
     * @return {?}
     */
    ExternalAlfrescoApiService.prototype.getInstance = /**
     * @return {?}
     */
    function () {
        return this.alfrescoApi;
    };
    Object.defineProperty(ExternalAlfrescoApiService.prototype, "contentApi", {
        get: /**
         * @return {?}
         */
        function () {
            return this.getInstance().content;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExternalAlfrescoApiService.prototype, "nodesApi", {
        get: /**
         * @return {?}
         */
        function () {
            return this.getInstance().nodes;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} ecmHost
     * @param {?} contextRoot
     * @return {?}
     */
    ExternalAlfrescoApiService.prototype.init = /**
     * @param {?} ecmHost
     * @param {?} contextRoot
     * @return {?}
     */
    function (ecmHost, contextRoot) {
        /** @type {?} */
        var domainPrefix = this.createPrefixFromHost(ecmHost);
        /** @type {?} */
        var config = {
            provider: 'ECM',
            hostEcm: ecmHost,
            authType: 'BASIC',
            contextRoot: contextRoot,
            domainPrefix: domainPrefix
        };
        this.initAlfrescoApi(config);
    };
    /**
     * @protected
     * @param {?} config
     * @return {?}
     */
    ExternalAlfrescoApiService.prototype.initAlfrescoApi = /**
     * @protected
     * @param {?} config
     * @return {?}
     */
    function (config) {
        if (this.alfrescoApi) {
            this.alfrescoApi.configureJsApi(config);
        }
        else {
            this.alfrescoApi = new AlfrescoApiCompatibility(config);
        }
    };
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    ExternalAlfrescoApiService.prototype.createPrefixFromHost = /**
     * @private
     * @param {?} url
     * @return {?}
     */
    function (url) {
        /** @type {?} */
        var match = url.match(/:\/\/(www[0-9]?\.)?(.[^/:]+)/i);
        /** @type {?} */
        var result = null;
        if (match != null && match.length > 2 && typeof match[2] === 'string' && match[2].length > 0) {
            result = match[2];
        }
        return result;
    };
    ExternalAlfrescoApiService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ExternalAlfrescoApiService.ngInjectableDef = i0.defineInjectable({ factory: function ExternalAlfrescoApiService_Factory() { return new ExternalAlfrescoApiService(); }, token: ExternalAlfrescoApiService, providedIn: "root" });
    return ExternalAlfrescoApiService;
}());
export { ExternalAlfrescoApiService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    ExternalAlfrescoApiService.prototype.alfrescoApi;
}
