// @alfresco/adf-core/esm2015/card-view/services/card-view-update.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function UpdateNotification() { }
if (false) {
    /** @type {?} */
    UpdateNotification.prototype.target;
    /** @type {?} */
    UpdateNotification.prototype.changed;
}
/**
 * @record
 */
export function ClickNotification() { }
if (false) {
    /** @type {?} */
    ClickNotification.prototype.target;
}
/**
 * @param {?} key
 * @param {?} value
 * @return {?}
 */
export function transformKeyToObject(key, value) {
    /** @type {?} */
    const objectLevels = key.split('.').reverse();
    return objectLevels.reduce((previousValue, currentValue) => {
        return { [currentValue]: previousValue };
    }, value);
}
export class CardViewUpdateService {
    constructor() {
        // Observable sources
        this.itemUpdatedSource = new Subject();
        this.itemClickedSource = new Subject();
        // Observable streams
        this.itemUpdated$ = (/** @type {?} */ (this.itemUpdatedSource.asObservable()));
        this.itemClicked$ = (/** @type {?} */ (this.itemClickedSource.asObservable()));
    }
    /**
     * @param {?} property
     * @param {?} newValue
     * @return {?}
     */
    update(property, newValue) {
        this.itemUpdatedSource.next({
            target: property,
            changed: transformKeyToObject(property.key, newValue)
        });
    }
    /**
     * @param {?} property
     * @return {?}
     */
    clicked(property) {
        this.itemClickedSource.next({
            target: property
        });
    }
}
CardViewUpdateService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ CardViewUpdateService.ngInjectableDef = i0.defineInjectable({ factory: function CardViewUpdateService_Factory() { return new CardViewUpdateService(); }, token: CardViewUpdateService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    CardViewUpdateService.prototype.itemUpdatedSource;
    /**
     * @type {?}
     * @private
     */
    CardViewUpdateService.prototype.itemClickedSource;
    /** @type {?} */
    CardViewUpdateService.prototype.itemUpdated$;
    /** @type {?} */
    CardViewUpdateService.prototype.itemClicked$;
}
