// @alfresco/adf-insights/esm2015/diagram/components/raphael/raphael-rhombus.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { Point } from './models/point';
import { RaphaelBase } from './raphael-base';
import { RaphaelService } from './raphael.service';
/**
 * Directive selectors without adf- prefix will be deprecated on 3.0.0
 */
export class RaphaelRhombusDirective extends RaphaelBase {
    /**
     * @param {?} elementRef
     * @param {?} raphaelService
     */
    constructor(elementRef, raphaelService) {
        super(elementRef, raphaelService);
        this.elementRef = elementRef;
        this.error = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        let opts = { 'stroke-width': this.strokeWidth, 'fill': this.fillColors, 'stroke': this.stroke, 'fill-opacity': this.fillOpacity };
        /** @type {?} */
        let elementDraw = this.draw(this.center, this.width, this.height, opts);
        elementDraw.node.id = this.elementId;
    }
    /**
     * @param {?} center
     * @param {?} width
     * @param {?} height
     * @param {?=} opts
     * @return {?}
     */
    draw(center, width, height, opts) {
        return this.paper.path('M' + center.x + ' ' + (center.y + (height / 2)) +
            'L' + (center.x + (width / 2)) + ' ' + (center.y + height) +
            'L' + (center.x + width) + ' ' + (center.y + (height / 2)) +
            'L' + (center.x + (width / 2)) + ' ' + center.y + 'z').attr(opts);
    }
}
RaphaelRhombusDirective.decorators = [
    { type: Directive, args: [{ selector: 'adf-raphael-rhombus, raphael-rhombus' },] }
];
/** @nocollapse */
RaphaelRhombusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: RaphaelService }
];
RaphaelRhombusDirective.propDecorators = {
    center: [{ type: Input }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    fillColors: [{ type: Input }],
    stroke: [{ type: Input }],
    strokeWidth: [{ type: Input }],
    fillOpacity: [{ type: Input }],
    elementId: [{ type: Input }],
    error: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    RaphaelRhombusDirective.prototype.center;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.width;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.height;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.fillColors;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.stroke;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.strokeWidth;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.fillOpacity;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.elementId;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.error;
    /** @type {?} */
    RaphaelRhombusDirective.prototype.elementRef;
}
