// @alfresco/adf-process-services/esm5/content-widget/attach-file-widget-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { MatDialog } from '@angular/material';
import { EventEmitter, Injectable, Output } from '@angular/core';
import { Subject } from 'rxjs';
import { AttachFileWidgetDialogComponent } from './attach-file-widget-dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/dialog";
var AttachFileWidgetDialogService = /** @class */ (function () {
    function AttachFileWidgetDialogService(dialog) {
        this.dialog = dialog;
        /**
         * Emitted when an error occurs.
         */
        this.error = new EventEmitter();
    }
    /**
     * Opens a dialog to choose a file to upload.
     * @param action Name of the action to show in the title
     * @param contentEntry Item to upload
     * @returns Information about the chosen file(s)
     */
    /**
     * Opens a dialog to choose a file to upload.
     * @param {?} ecmHost
     * @param {?=} actionName
     * @param {?=} context
     * @return {?} Information about the chosen file(s)
     */
    AttachFileWidgetDialogService.prototype.openLogin = /**
     * Opens a dialog to choose a file to upload.
     * @param {?} ecmHost
     * @param {?=} actionName
     * @param {?=} context
     * @return {?} Information about the chosen file(s)
     */
    function (ecmHost, actionName, context) {
        /** @type {?} */
        var titleString = "Please log in for " + ecmHost;
        /** @type {?} */
        var selected = new Subject();
        selected.subscribe({
            complete: this.close.bind(this)
        });
        /** @type {?} */
        var data = {
            title: titleString,
            actionName: actionName,
            selected: selected,
            ecmHost: ecmHost,
            context: context,
            isSelectionValid: this.isNodeFile.bind(this)
        };
        this.openLoginDialog(data, 'adf-attach-file-widget-dialog', '630px');
        return selected;
    };
    /**
     * @private
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    AttachFileWidgetDialogService.prototype.openLoginDialog = /**
     * @private
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    function (data, currentPanelClass, chosenWidth) {
        this.dialog.open(AttachFileWidgetDialogComponent, { data: data, panelClass: currentPanelClass, width: chosenWidth });
    };
    /** Closes the currently open dialog. */
    /**
     * Closes the currently open dialog.
     * @return {?}
     */
    AttachFileWidgetDialogService.prototype.close = /**
     * Closes the currently open dialog.
     * @return {?}
     */
    function () {
        this.dialog.closeAll();
    };
    /**
     * @private
     * @param {?} entry
     * @return {?}
     */
    AttachFileWidgetDialogService.prototype.isNodeFile = /**
     * @private
     * @param {?} entry
     * @return {?}
     */
    function (entry) {
        return entry.isFile;
    };
    AttachFileWidgetDialogService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AttachFileWidgetDialogService.ctorParameters = function () { return [
        { type: MatDialog }
    ]; };
    AttachFileWidgetDialogService.propDecorators = {
        error: [{ type: Output }]
    };
    /** @nocollapse */ AttachFileWidgetDialogService.ngInjectableDef = i0.defineInjectable({ factory: function AttachFileWidgetDialogService_Factory() { return new AttachFileWidgetDialogService(i0.inject(i1.MatDialog)); }, token: AttachFileWidgetDialogService, providedIn: "root" });
    return AttachFileWidgetDialogService;
}());
export { AttachFileWidgetDialogService };
if (false) {
    /**
     * Emitted when an error occurs.
     * @type {?}
     */
    AttachFileWidgetDialogService.prototype.error;
    /**
     * @type {?}
     * @private
     */
    AttachFileWidgetDialogService.prototype.dialog;
}
