// @alfresco/adf-core/esm5/directives/logout.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Input, Directive, ElementRef, Renderer2 } from '@angular/core';
import { Router } from '@angular/router';
import { AuthenticationService } from '../services/authentication.service';
var LogoutDirective = /** @class */ (function () {
    function LogoutDirective(elementRef, renderer, router, auth) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.router = router;
        this.auth = auth;
        /**
         * URI to redirect to after logging out.
         */
        this.redirectUri = '/login';
        /**
         * Enable redirecting after logout
         */
        this.enableRedirect = true;
    }
    /**
     * @return {?}
     */
    LogoutDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.elementRef.nativeElement) {
            this.renderer.listen(this.elementRef.nativeElement, 'click', function (evt) {
                evt.preventDefault();
                _this.logout();
            });
        }
    };
    /**
     * @return {?}
     */
    LogoutDirective.prototype.logout = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.auth.logout().subscribe(function () { return _this.redirectToUri(); }, function () { return _this.redirectToUri(); });
    };
    /**
     * @return {?}
     */
    LogoutDirective.prototype.redirectToUri = /**
     * @return {?}
     */
    function () {
        if (this.enableRedirect) {
            this.router.navigate([this.redirectUri]);
        }
    };
    LogoutDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[adf-logout]'
                },] }
    ];
    /** @nocollapse */
    LogoutDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Router },
        { type: AuthenticationService }
    ]; };
    LogoutDirective.propDecorators = {
        redirectUri: [{ type: Input }],
        enableRedirect: [{ type: Input }]
    };
    return LogoutDirective;
}());
export { LogoutDirective };
if (false) {
    /**
     * URI to redirect to after logging out.
     * @type {?}
     */
    LogoutDirective.prototype.redirectUri;
    /**
     * Enable redirecting after logout
     * @type {?}
     */
    LogoutDirective.prototype.enableRedirect;
    /**
     * @type {?}
     * @private
     */
    LogoutDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    LogoutDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    LogoutDirective.prototype.router;
    /**
     * @type {?}
     * @private
     */
    LogoutDirective.prototype.auth;
}
