// @alfresco/adf-content-services/esm2015/document-list/components/library-role-column/library-role-column.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Input } from '@angular/core';
export class LibraryRoleColumnComponent {
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const node = this.context.row.node;
        if (node && node.entry) {
            /** @type {?} */
            const role = node.entry.role;
            switch (role) {
                case 'SiteManager':
                    this.displayText = 'LIBRARY.ROLE.MANAGER';
                    break;
                case 'SiteCollaborator':
                    this.displayText = 'LIBRARY.ROLE.COLLABORATOR';
                    break;
                case 'SiteContributor':
                    this.displayText = 'LIBRARY.ROLE.CONTRIBUTOR';
                    break;
                case 'SiteConsumer':
                    this.displayText = 'LIBRARY.ROLE.CONSUMER';
                    break;
                default:
                    this.displayText = '';
                    break;
            }
        }
    }
}
LibraryRoleColumnComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-library-role-column',
                template: `
    <span title="{{ displayText | translate }}">
      {{ displayText | translate }}
    </span>
  `,
                host: { class: 'adf-library-role-column' }
            }] }
];
LibraryRoleColumnComponent.propDecorators = {
    context: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    LibraryRoleColumnComponent.prototype.context;
    /** @type {?} */
    LibraryRoleColumnComponent.prototype.displayText;
}
