// @alfresco/adf-core/esm2015/services/auth-guard-ecm.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AuthenticationService } from './authentication.service';
import { AppConfigService, AppConfigValues } from '../app-config/app-config.service';
import * as i0 from "@angular/core";
import * as i1 from "./authentication.service";
import * as i2 from "@angular/router";
import * as i3 from "../app-config/app-config.service";
export class AuthGuardEcm {
    /**
     * @param {?} authService
     * @param {?} router
     * @param {?} appConfigService
     */
    constructor(authService, router, appConfigService) {
        this.authService = authService;
        this.router = router;
        this.appConfigService = appConfigService;
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    canActivate(route, state) {
        return this.checkLogin(state.url);
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    canActivateChild(route, state) {
        return this.canActivate(route, state);
    }
    /**
     * @param {?} redirectUrl
     * @return {?}
     */
    checkLogin(redirectUrl) {
        /** @type {?} */
        let withCredentialsMode = this.appConfigService.get('auth.withCredentials', false);
        if (this.authService.isEcmLoggedIn() || withCredentialsMode) {
            return true;
        }
        if (!this.authService.isOauth() || this.isOAuthWithoutSilentLogin()) {
            this.authService.setRedirect({ provider: 'ECM', url: redirectUrl });
            /** @type {?} */
            const pathToLogin = this.getRouteDestinationForLogin();
            this.router.navigate(['/' + pathToLogin]);
        }
        return false;
    }
    /**
     * @return {?}
     */
    isOAuthWithoutSilentLogin() {
        /** @type {?} */
        let oauth = this.appConfigService.get(AppConfigValues.OAUTHCONFIG, null);
        return this.authService.isOauth() && oauth.silentLogin === false;
    }
    /**
     * @private
     * @return {?}
     */
    getRouteDestinationForLogin() {
        return this.appConfigService &&
            this.appConfigService.get(AppConfigValues.LOGIN_ROUTE) ?
            this.appConfigService.get(AppConfigValues.LOGIN_ROUTE) : 'login';
    }
}
AuthGuardEcm.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AuthGuardEcm.ctorParameters = () => [
    { type: AuthenticationService },
    { type: Router },
    { type: AppConfigService }
];
/** @nocollapse */ AuthGuardEcm.ngInjectableDef = i0.defineInjectable({ factory: function AuthGuardEcm_Factory() { return new AuthGuardEcm(i0.inject(i1.AuthenticationService), i0.inject(i2.Router), i0.inject(i3.AppConfigService)); }, token: AuthGuardEcm, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthGuardEcm.prototype.authService;
    /**
     * @type {?}
     * @private
     */
    AuthGuardEcm.prototype.router;
    /**
     * @type {?}
     * @private
     */
    AuthGuardEcm.prototype.appConfigService;
}
