// @alfresco/adf-content-services/esm5/document-list/components/library-status-column/library-status-column.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Input } from '@angular/core';
var LibraryStatusColumnComponent = /** @class */ (function () {
    function LibraryStatusColumnComponent() {
    }
    /**
     * @return {?}
     */
    LibraryStatusColumnComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var node = this.context.row.node;
        if (node && node.entry) {
            /** @type {?} */
            var visibility = node.entry.visibility;
            switch (visibility.toUpperCase()) {
                case 'PUBLIC':
                    this.displayText = 'LIBRARY.VISIBILITY.PUBLIC';
                    break;
                case 'PRIVATE':
                    this.displayText = 'LIBRARY.VISIBILITY.PRIVATE';
                    break;
                case 'MODERATED':
                    this.displayText = 'LIBRARY.VISIBILITY.MODERATED';
                    break;
                default:
                    this.displayText = 'UNKNOWN';
                    break;
            }
        }
    };
    LibraryStatusColumnComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-library-status-column',
                    template: "\n    <span title=\"{{ displayText | translate }}\">\n      {{ displayText | translate }}\n    </span>\n  ",
                    host: { class: 'adf-library-status-column' }
                }] }
    ];
    LibraryStatusColumnComponent.propDecorators = {
        context: [{ type: Input }]
    };
    return LibraryStatusColumnComponent;
}());
export { LibraryStatusColumnComponent };
if (false) {
    /** @type {?} */
    LibraryStatusColumnComponent.prototype.context;
    /** @type {?} */
    LibraryStatusColumnComponent.prototype.displayText;
}
