// @alfresco/adf-core/esm5/datatable/components/datatable/datatable-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
var DataTableCellComponent = /** @class */ (function () {
    function DataTableCellComponent() {
    }
    /**
     * @return {?}
     */
    DataTableCellComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.value && this.column && this.column.key && this.row && this.data) {
            this.value = this.data.getValue(this.row, this.column);
            if (!this.tooltip) {
                this.tooltip = this.value;
            }
        }
    };
    DataTableCellComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-datatable-cell',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n        <ng-container>\n            <span [title]=\"tooltip\" class=\"adf-datatable-cell-value\">{{value}}</span>\n        </ng-container>",
                    encapsulation: ViewEncapsulation.None,
                    host: { class: 'adf-datatable-cell' }
                }] }
    ];
    DataTableCellComponent.propDecorators = {
        data: [{ type: Input }],
        column: [{ type: Input }],
        row: [{ type: Input }],
        value: [{ type: Input }],
        tooltip: [{ type: Input }]
    };
    return DataTableCellComponent;
}());
export { DataTableCellComponent };
if (false) {
    /** @type {?} */
    DataTableCellComponent.prototype.data;
    /** @type {?} */
    DataTableCellComponent.prototype.column;
    /** @type {?} */
    DataTableCellComponent.prototype.row;
    /** @type {?} */
    DataTableCellComponent.prototype.value;
    /** @type {?} */
    DataTableCellComponent.prototype.tooltip;
}
