// @alfresco/adf-process-services/esm5/task-list/components/task-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { DataTableSchema, CustomEmptyContentTemplateDirective, CustomLoadingContentTemplateDirective } from '@alfresco/adf-core';
import { AppConfigService, PaginationComponent, UserPreferencesService, UserPreferenceValues } from '@alfresco/adf-core';
import { Component, ContentChild, EventEmitter, Input, Output } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { TaskQueryRequestRepresentationModel } from '../models/filter.model';
import { taskPresetsDefaultModel } from '../models/task-preset.model';
import { TaskListService } from './../services/tasklist.service';
import moment from 'moment-es6';
var TaskListComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaskListComponent, _super);
    function TaskListComponent(taskListService, appConfigService, userPreferences) {
        var _this = _super.call(this, appConfigService, TaskListComponent.PRESET_KEY, taskPresetsDefaultModel) || this;
        _this.taskListService = taskListService;
        _this.userPreferences = userPreferences;
        _this.FORMAT_DATE = 'll';
        /**
         * Row selection mode. Can be none, `single` or `multiple`. For `multiple` mode,
         * you can use Cmd (macOS) or Ctrl (Win) modifier key to toggle selection for
         * multiple rows.
         */
        _this.selectionMode = 'single'; // none|single|multiple
        // none|single|multiple
        /**
         * Toggles multiple row selection, renders checkboxes at the beginning of each row
         */
        _this.multiselect = false;
        /**
         * Toggles default selection of the first row
         */
        _this.selectFirstRow = true;
        /**
         * Emitted when a task in the list is clicked
         */
        _this.rowClick = new EventEmitter();
        /**
         * Emitted when rows are selected/unselected
         */
        _this.rowsSelected = new EventEmitter();
        /**
         * Emitted when the task list is loaded
         */
        _this.success = new EventEmitter();
        /**
         * Emitted when an error occurs.
         */
        _this.error = new EventEmitter();
        /**
         * The page number of the tasks to fetch.
         */
        _this.page = 0;
        /**
         * The number of tasks to fetch. Default value: 25.
         */
        _this.size = PaginationComponent.DEFAULT_PAGINATION.maxItems;
        _this.rows = [];
        _this.isLoading = true;
        _this.sorting = ['created', 'desc'];
        /**
         * Toggles custom data source mode.
         * When enabled the component reloads data from it's current source instead of the server side.
         * This allows generating and displaying custom data sets (i.e. filtered out content).
         *
         * \@memberOf TaskListComponent
         */
        _this.hasCustomDataSource = false;
        _this.userPreferences.select(UserPreferenceValues.PaginationSize).subscribe(function (pageSize) {
            _this.size = pageSize;
        });
        _this.pagination = new BehaviorSubject((/** @type {?} */ ({
            maxItems: _this.size,
            skipCount: 0,
            totalItems: 0
        })));
        return _this;
    }
    /**
     * @return {?}
     */
    TaskListComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.createDatatableSchema();
        if (this.data && this.data.getColumns().length === 0) {
            this.data.setColumns(this.columns);
        }
        if (this.appId) {
            this.reload();
        }
    };
    /**
     * @param {?} rows
     * @return {?}
     */
    TaskListComponent.prototype.setCustomDataSource = /**
     * @param {?} rows
     * @return {?}
     */
    function (rows) {
        if (rows) {
            this.rows = rows;
            this.hasCustomDataSource = true;
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    TaskListComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.isPropertyChanged(changes)) {
            if (this.isSortChanged(changes)) {
                this.sorting = this.sort ? this.sort.split('-') : this.sorting;
            }
            this.reload();
        }
    };
    /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    TaskListComponent.prototype.isSortChanged = /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var actualSort = changes['sort'];
        return actualSort && actualSort.currentValue && actualSort.currentValue !== actualSort.previousValue;
    };
    /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    TaskListComponent.prototype.isPropertyChanged = /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var changed = true;
        /** @type {?} */
        var landingTaskId = changes['landingTaskId'];
        /** @type {?} */
        var page = changes['page'];
        /** @type {?} */
        var size = changes['size'];
        if (landingTaskId && landingTaskId.currentValue && this.isEqualToCurrentId(landingTaskId.currentValue)) {
            changed = false;
        }
        else if (page && page.currentValue !== page.previousValue) {
            changed = true;
        }
        else if (size && size.currentValue !== size.previousValue) {
            changed = true;
        }
        return changed;
    };
    /**
     * @return {?}
     */
    TaskListComponent.prototype.reload = /**
     * @return {?}
     */
    function () {
        if (!this.hasCustomDataSource) {
            this.requestNode = this.createRequestNode();
            this.load(this.requestNode);
        }
        else {
            this.isLoading = false;
        }
    };
    /**
     * @private
     * @param {?} requestNode
     * @return {?}
     */
    TaskListComponent.prototype.load = /**
     * @private
     * @param {?} requestNode
     * @return {?}
     */
    function (requestNode) {
        var _this = this;
        this.isLoading = true;
        this.loadTasksByState().subscribe(function (tasks) {
            _this.rows = _this.optimizeTaskDetails(tasks.data);
            _this.selectTask(_this.landingTaskId);
            _this.success.emit(tasks);
            _this.isLoading = false;
            _this.pagination.next({
                count: tasks.data.length,
                maxItems: _this.size,
                skipCount: _this.page * _this.size,
                totalItems: tasks.total
            });
        }, function (error) {
            _this.error.emit(error);
            _this.isLoading = false;
        });
    };
    /**
     * @private
     * @return {?}
     */
    TaskListComponent.prototype.loadTasksByState = /**
     * @private
     * @return {?}
     */
    function () {
        return this.requestNode.state === 'all'
            ? this.taskListService.findAllTasksWithoutState(this.requestNode)
            : this.taskListService.findTasksByState(this.requestNode);
    };
    /**
     * Select the task given in input if present
     */
    /**
     * Select the task given in input if present
     * @param {?} taskIdSelected
     * @return {?}
     */
    TaskListComponent.prototype.selectTask = /**
     * Select the task given in input if present
     * @param {?} taskIdSelected
     * @return {?}
     */
    function (taskIdSelected) {
        if (!this.isListEmpty()) {
            /** @type {?} */
            var dataRow = null;
            if (taskIdSelected) {
                dataRow = this.rows.find(function (currentRow) {
                    return currentRow['id'] === taskIdSelected;
                });
            }
            if (!dataRow && this.selectFirstRow) {
                dataRow = this.rows[0];
            }
            if (dataRow) {
                dataRow.isSelected = true;
                this.currentInstanceId = dataRow['id'];
            }
        }
        else {
            this.currentInstanceId = null;
        }
    };
    /**
     * Return the current id
     */
    /**
     * Return the current id
     * @return {?}
     */
    TaskListComponent.prototype.getCurrentId = /**
     * Return the current id
     * @return {?}
     */
    function () {
        return this.currentInstanceId;
    };
    /**
     * Check if the taskId is the same of the selected task
     * @param taskId
     */
    /**
     * Check if the taskId is the same of the selected task
     * @param {?} taskId
     * @return {?}
     */
    TaskListComponent.prototype.isEqualToCurrentId = /**
     * Check if the taskId is the same of the selected task
     * @param {?} taskId
     * @return {?}
     */
    function (taskId) {
        return this.currentInstanceId === taskId;
    };
    /**
     * Check if the list is empty
     */
    /**
     * Check if the list is empty
     * @return {?}
     */
    TaskListComponent.prototype.isListEmpty = /**
     * Check if the list is empty
     * @return {?}
     */
    function () {
        return !this.rows || this.rows.length === 0;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    TaskListComponent.prototype.onRowClick = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.currentInstanceId = item.value.getValue('id');
        this.rowClick.emit(this.currentInstanceId);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TaskListComponent.prototype.onRowSelect = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.selectedInstances = tslib_1.__spread(event.detail.selection);
        this.rowsSelected.emit(this.selectedInstances);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TaskListComponent.prototype.onRowUnselect = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.selectedInstances = tslib_1.__spread(event.detail.selection);
        this.rowsSelected.emit(this.selectedInstances);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TaskListComponent.prototype.onRowKeyUp = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.detail.keyboardEvent.key === 'Enter') {
            event.preventDefault();
            this.currentInstanceId = event.detail.row.getValue('id');
            this.rowClick.emit(this.currentInstanceId);
        }
    };
    /**
     * Optimize name field
     * @param instances
     */
    /**
     * Optimize name field
     * @private
     * @param {?} instances
     * @return {?}
     */
    TaskListComponent.prototype.optimizeTaskDetails = /**
     * Optimize name field
     * @private
     * @param {?} instances
     * @return {?}
     */
    function (instances) {
        var _this = this;
        instances = instances.map(function (task) {
            if (!task.name) {
                task.name = 'No name';
            }
            if (task.created) {
                task.created = moment(task.created).format(_this.FORMAT_DATE);
            }
            if (task.dueDate) {
                task.dueDate = moment(task.dueDate).format(_this.FORMAT_DATE);
            }
            if (task.endDate) {
                task.endDate = moment(task.endDate).format(_this.FORMAT_DATE);
            }
            return task;
        });
        return instances;
    };
    /**
     * @private
     * @return {?}
     */
    TaskListComponent.prototype.createRequestNode = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var requestNode = {
            appDefinitionId: this.appId,
            dueAfter: this.dueAfter ? moment(this.dueAfter).toDate() : null,
            dueBefore: this.dueBefore ? moment(this.dueBefore).toDate() : null,
            processInstanceId: this.processInstanceId,
            processDefinitionId: this.processDefinitionId,
            text: this.name,
            assignment: this.assignment,
            state: this.state,
            sort: this.sort,
            page: this.page,
            size: this.size,
            start: this.start,
            taskId: this.taskId,
            includeProcessInstance: this.includeProcessInstance
        };
        return new TaskQueryRequestRepresentationModel(requestNode);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    TaskListComponent.prototype.updatePagination = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        /** @type {?} */
        var needsReload = params.maxItems || params.skipCount;
        this.size = params.maxItems;
        this.page = this.currentPage(params.skipCount, params.maxItems);
        if (needsReload) {
            this.reload();
        }
    };
    /**
     * @param {?} skipCount
     * @param {?} maxItems
     * @return {?}
     */
    TaskListComponent.prototype.currentPage = /**
     * @param {?} skipCount
     * @param {?} maxItems
     * @return {?}
     */
    function (skipCount, maxItems) {
        return (skipCount && maxItems) ? Math.floor(skipCount / maxItems) : 0;
    };
    TaskListComponent.PRESET_KEY = 'adf-task-list.presets';
    TaskListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-tasklist',
                    template: "<div *ngIf=\"!requestNode\">{{ 'ADF_TASK_LIST.FILTERS.MESSAGES.NONE' | translate }}</div>\n<ng-container *ngIf=\"requestNode\">\n        <adf-datatable\n            [data]=\"data\"\n            [rows]=\"rows\"\n            [columns]=\"columns\"\n            [sorting]=\"sorting\"\n            [loading]=\"isLoading\"\n            [multiselect]=\"multiselect\"\n            [selectionMode]=\"selectionMode\"\n            (row-select)=\"onRowSelect($event)\"\n            (row-unselect)=\"onRowUnselect($event)\"\n            (rowClick)=\"onRowClick($event)\"\n            (row-keyup)=\"onRowKeyUp($event)\">\n            <adf-loading-content-template>\n                <ng-template>\n                    <!--Add your custom loading template here-->\n                    <mat-progress-spinner\n                        *ngIf=\"!customLoadingContent\"\n                        class=\"adf-task-list-loading-margin\"\n                        [color]=\"'primary'\"\n                        [mode]=\"'indeterminate'\">\n                    </mat-progress-spinner>\n                    <ng-content select=\"adf-custom-loading-content-template\"></ng-content>\n                </ng-template>\n            </adf-loading-content-template>\n            <adf-no-content-template>\n                <ng-template>\n                    <adf-empty-content *ngIf=\"!customEmptyContent\"\n                        icon=\"assignment\"\n                        [title]=\"'ADF_TASK_LIST.LIST.MESSAGES.TITLE' | translate\"\n                        [subtitle]=\"'ADF_TASK_LIST.LIST.MESSAGES.SUBTITLE' | translate\">\n                    </adf-empty-content>\n                    <ng-content select=\"adf-custom-empty-content-template\"></ng-content>\n                </ng-template>\n            </adf-no-content-template>\n        </adf-datatable>\n</ng-container>\n",
                    styles: [".adf-task-list-loading-margin{margin-left:calc((100% - 100px)/ 2);margin-right:calc((100% - 100px)/ 2)}"]
                }] }
    ];
    /** @nocollapse */
    TaskListComponent.ctorParameters = function () { return [
        { type: TaskListService },
        { type: AppConfigService },
        { type: UserPreferencesService }
    ]; };
    TaskListComponent.propDecorators = {
        customEmptyContent: [{ type: ContentChild, args: [CustomEmptyContentTemplateDirective,] }],
        customLoadingContent: [{ type: ContentChild, args: [CustomLoadingContentTemplateDirective,] }],
        appId: [{ type: Input }],
        processInstanceId: [{ type: Input }],
        processDefinitionId: [{ type: Input }],
        state: [{ type: Input }],
        assignment: [{ type: Input }],
        sort: [{ type: Input }],
        name: [{ type: Input }],
        landingTaskId: [{ type: Input }],
        data: [{ type: Input }],
        selectionMode: [{ type: Input }],
        multiselect: [{ type: Input }],
        selectFirstRow: [{ type: Input }],
        taskId: [{ type: Input }],
        includeProcessInstance: [{ type: Input }],
        start: [{ type: Input }],
        rowClick: [{ type: Output }],
        rowsSelected: [{ type: Output }],
        success: [{ type: Output }],
        error: [{ type: Output }],
        page: [{ type: Input }],
        size: [{ type: Input }],
        dueAfter: [{ type: Input }],
        dueBefore: [{ type: Input }]
    };
    return TaskListComponent;
}(DataTableSchema));
export { TaskListComponent };
if (false) {
    /** @type {?} */
    TaskListComponent.PRESET_KEY;
    /** @type {?} */
    TaskListComponent.prototype.FORMAT_DATE;
    /** @type {?} */
    TaskListComponent.prototype.customEmptyContent;
    /** @type {?} */
    TaskListComponent.prototype.customLoadingContent;
    /** @type {?} */
    TaskListComponent.prototype.requestNode;
    /**
     * The id of the app.
     * @type {?}
     */
    TaskListComponent.prototype.appId;
    /**
     * The Instance Id of the process.
     * @type {?}
     */
    TaskListComponent.prototype.processInstanceId;
    /**
     * The Definition Id of the process.
     * @type {?}
     */
    TaskListComponent.prototype.processDefinitionId;
    /**
     * Current state of the process. Possible values are: `completed`, `active`.
     * @type {?}
     */
    TaskListComponent.prototype.state;
    /**
     * The assignment of the process. Possible values are: "assignee" (the current user
     * is the assignee), "candidate" (the current user is a task candidate, "group_x" (the task
     * is assigned to a group where the current user is a member,
     * no value (the current user is involved).
     * @type {?}
     */
    TaskListComponent.prototype.assignment;
    /**
     * Define the sort order of the tasks. Possible values are : `created-desc`,
     * `created-asc`, `due-desc`, `due-asc`
     * @type {?}
     */
    TaskListComponent.prototype.sort;
    /**
     * Name of the tasklist.
     * @type {?}
     */
    TaskListComponent.prototype.name;
    /**
     * Define which task id should be selected after reloading. If the task id doesn't
     * exist or nothing is passed then the first task will be selected.
     * @type {?}
     */
    TaskListComponent.prototype.landingTaskId;
    /**
     * Data source object that represents the number and the type of the columns that
     * you want to show.
     * @type {?}
     */
    TaskListComponent.prototype.data;
    /**
     * Row selection mode. Can be none, `single` or `multiple`. For `multiple` mode,
     * you can use Cmd (macOS) or Ctrl (Win) modifier key to toggle selection for
     * multiple rows.
     * @type {?}
     */
    TaskListComponent.prototype.selectionMode;
    /**
     * Toggles multiple row selection, renders checkboxes at the beginning of each row
     * @type {?}
     */
    TaskListComponent.prototype.multiselect;
    /**
     * Toggles default selection of the first row
     * @type {?}
     */
    TaskListComponent.prototype.selectFirstRow;
    /**
     * The id of a task
     * @type {?}
     */
    TaskListComponent.prototype.taskId;
    /**
     * Toggles inclusion of Process Instances
     * @type {?}
     */
    TaskListComponent.prototype.includeProcessInstance;
    /**
     * Starting point of the list within the full set of tasks.
     * @type {?}
     */
    TaskListComponent.prototype.start;
    /**
     * Emitted when a task in the list is clicked
     * @type {?}
     */
    TaskListComponent.prototype.rowClick;
    /**
     * Emitted when rows are selected/unselected
     * @type {?}
     */
    TaskListComponent.prototype.rowsSelected;
    /**
     * Emitted when the task list is loaded
     * @type {?}
     */
    TaskListComponent.prototype.success;
    /**
     * Emitted when an error occurs.
     * @type {?}
     */
    TaskListComponent.prototype.error;
    /** @type {?} */
    TaskListComponent.prototype.currentInstanceId;
    /** @type {?} */
    TaskListComponent.prototype.selectedInstances;
    /** @type {?} */
    TaskListComponent.prototype.pagination;
    /**
     * The page number of the tasks to fetch.
     * @type {?}
     */
    TaskListComponent.prototype.page;
    /**
     * The number of tasks to fetch. Default value: 25.
     * @type {?}
     */
    TaskListComponent.prototype.size;
    /**
     * Filter the tasks. Display only tasks with `created_date` after `dueAfter`.
     * @type {?}
     */
    TaskListComponent.prototype.dueAfter;
    /**
     * Filter the tasks. Display only tasks with `created_date` before `dueBefore`.
     * @type {?}
     */
    TaskListComponent.prototype.dueBefore;
    /** @type {?} */
    TaskListComponent.prototype.rows;
    /** @type {?} */
    TaskListComponent.prototype.isLoading;
    /** @type {?} */
    TaskListComponent.prototype.sorting;
    /**
     * Toggles custom data source mode.
     * When enabled the component reloads data from it's current source instead of the server side.
     * This allows generating and displaying custom data sets (i.e. filtered out content).
     *
     * \@memberOf TaskListComponent
     * @type {?}
     */
    TaskListComponent.prototype.hasCustomDataSource;
    /**
     * @type {?}
     * @private
     */
    TaskListComponent.prototype.taskListService;
    /**
     * @type {?}
     * @private
     */
    TaskListComponent.prototype.userPreferences;
}
